src/utils/mod.rs

```diff
@@ -246,30 +250,77 @@ async fn download_file_(
     // Download the file
 
     // Keep the curl env var around for a bit
-    let use_curl_backend = process
-        .var_os("RUSTUP_USE_CURL")
-        .is_some_and(|it| it != "0");
-    let use_rustls = process
-        .var_os("RUSTUP_USE_RUSTLS")
-        .is_none_or(|it| it != "0");
-    let (backend, notification) = if use_curl_backend {
-        (Backend::Curl, Notification::UsingCurl)
-    } else {
-        let tls_backend = if use_rustls {
-            TlsBackend::Rustls
-        } else {
+    let use_curl_backend = process.var_os("RUSTUP_USE_CURL").map(|it| it != "0");
+    #[cfg(not(feature = "curl-backend"))]
+    if use_curl_backend == Some(true) {
+        return Err(anyhow!(
+            "RUSTUP_USE_CURL is set, but this rustup distribution was not built with the curl-backend feature"
+        ));
+    }
+
+    let use_rustls = process.var_os("RUSTUP_USE_RUSTLS").map(|it| it != "0");
+    #[cfg(not(feature = "reqwest-rustls-tls"))]
+    if use_rustls == Some(true) {
+        return Err(anyhow!(
+            "RUSTUP_USE_RUSTLS is set, but this rustup distribution was not built with the reqwest-rustls-tls feature"
+        ));
+    }
+    #[cfg(not(feature = "reqwest-native-tls"))]
+    if use_rustls == Some(false) {
+        return Err(anyhow!(
+            "RUSTUP_USE_RUSTLS is set to false, but this rustup distribution was not built with the reqwest-native-tls feature"
+        ));
+    }
+
+    let backend = match use_curl_backend {
+        #[cfg(feature = "reqwest-backend")]
+        None | Some(false) => Backend::Reqwest(match use_rustls {
+            #[cfg(feature = "reqwest-rustls-tls")]
+            Some(true) | None => {
+                if use_curl_backend == Some(true) {
```

Comment:
Looks like a typo here? You matched `use_curl_backend` to be `None | Some(false)` so it can't be `Some(true)`?

Reply:
You're right. So this is the reason I think a big `match` expression is actually a better approach, because it makes these cases more explicit.

Reply:
> You're right. So this is the reason I think a big `match` expression is actually a better approach, because it makes these cases more explicit.

@djc I'd say it's okay to keep the match as long as it's clear where every conditional branch comes from.

Reply:
> > You're right. So this is the reason I think a big `match` expression is actually a better approach, because it makes these cases more explicit.
> 
> @djc I'd say it's okay to keep the match as long as it's clear where every conditional branch comes from.

The `_` is actually load-bearing if you factor in variance due to Cargo features.

Reply:
@djc Thanks! This is looking much cleaner now.